### ISSUE TITLE
Last line of output from server

### ISSUE THREAD

**Issue by chrisharris333** [Spksh/TentacleSoftware.Telnet#6]
It seems that the last line of any message from the server does not trigger the MessageReceived event.  Is there some way to flush this out?  Example snippet:

`
_telnetClient = new TelnetClient("mybbs", 23, TimeSpan.FromSeconds(3), cToken);
_telnetClient.ConnectionClosed += HandleConnectionClosed;
_telnetClient.MessageReceived += HandleMessageReceived;
  await _telnetClient.Connect();

private void HandleMessageReceived(object sender, string message)
        {
            System.Windows.Forms.MessageBox.Show(message);
        }
`

In every case this code will not produce a messagebox for the last line of any transmission from any telnet server I have tried it on.

**Comment by Spksh** [comment]
@chrisharris333 can you provide a more complete repro, with a sample program, please? I hate to say "it works on my machine", but... 😄 

Also, the library is pretty chatty if you hook up a Trace listener. Do you get any interesting errors or warnings there?

**Comment by chrisharris333** [comment]
Thank you for your response.


I did figure out why I wasn't seeing what I was expecting.  You are using stream.ReadLine(), which requires some form of line feed/carriage return being sent from the telnet server.  The BBSs that I am connecting to do not always include such, for example, the password entry line says simply:

password:_

And the cursor is left immediately to the right of the password prompt.  Because there is no CRLF, \r\n, Enfironment.NewLine, etc. the line is never picked up by your code.

Refactoring would probably be more effort than it is worth, as I am sure the way you are doing things works fine for most telnet servers.  I am now using a different library.

Thanks again for the response.
-Chris Harris
mytinycloud.com




Oct 11, 2020, 11:18 PM by notifications@github.com:

**Comment by blakepell** [comment]
@chrisharris333 I had the same use case, I have a mud client (text based DnD game).  The server in my case doesn't always send a line feed and it was necessary for me to read the characters before the new line came in (I present them on the screen as they come in but would run triggers/RegEx's against them when the line feed actually comes).  As a result I modified this client with these changes:

1.) I have a `OnDataReceived` and a `OnLineReceived` event.  `OnDataReceived` is the data as soon as it gets there, `OnLineReceived` fires as it finds complete lines.
2.) Because the game I connect to sends lots of text (hundreds of thousands of lines if you play for a few hours) I made some changes to minimize (but not eliminate) string allocations (I'm re-using a `char[]`  and two `StringBuilder` classes).
3.) I added support for telnet go ahead (and it's probably not correct officially but it is correct for how most mud code bases use telnetga which is an indicator to the client to barf out whatever is in the buffer to the screen).
4.) I added a `IsConnected()` method so the client can poll to see if the connection has been broken (if the user is idle and not entering commands and the connection has been terminated the user won't know until they enter something, I use the `IsConnected` to poll the connection and re-connect which is important for the text game).  That portion might be a good fit for this library.

Here's a link to the modification I made, perhaps it will help you (not sure if it matters but I attribution and a link to this repo in my README).

https://github.com/blakepell/AvalonMudClient/blob/master/src/Avalon.Client/Network/TelnetClient.cs
